.github/workflows/edge_test.yml

```diff
@@ -137,3 +137,37 @@ jobs:
           OPENDAL_S3_BUCKET: opendal-testing
           OPENDAL_S3_ROLE_ARN: arn:aws:iam::952853449216:role/opendal-testing
           OPENDAL_S3_REGION: ap-northeast-1
+
+  test_s3_on_versionning_enable_bucket:
```

Comment:
This should be part of behavior test instead of edge test. We can add a new s3 setup with version enabled.

Reply:
It's an extra burden for other unrelated tests, and I think it's unnecessary. 🤔 
So I put this in the edge tests.
(Of course, a new s3 setup is necessary, wherever it is put.)

Reply:
> It's an extra burden for other unrelated tests, and I think it's unnecessary. 🤔

It's our public API that many services could implement. Adding edge tests for them one by one adds more burden.

Reply:
Is it possible for us to set up some additional composite actions for those cases where we need to add special settings to the service itself?
Agree that setting up one by one is cumbersome.

But for now, we could set it first.

Reply:
> Is it possible for us to set up some additional composite actions for those cases where we need to add special settings to the service itself?

It's an existing nature that some s3 bucket doesn't enable object versioning. So the steps we need to take is:

- Add read_with_version, stat_with_version capabilities.
- Add enable_version for s3 service. We will enable corresponding capabilities if this flag has been enabled.

Reply:
I'm mainly considering the cleanup issue. With storage services that have version control, cleanup can be more complicated. We'd have to remove every version of each file, so I hope to minimize the scope of the impact as much as possible. Since version control should not affect normal user usage, there should be no need to test the functionality of version control unless it is enabled.

So, can we:

1. Add version-related test cases to the existing behavior test. In this way, we can use environment variables to skip cases that do not involve version control, based on whether version control is enabled. But it also needs to add some extra cleanup steps.
1. Create a separate behavior test suite(maybe like this PR, but not in edge tests) that does not affect existing tests.

Reply:
> I'm mainly considering the cleanup issue. With storage services that have version control, cleanup can be more complicated.

Storage services have version support will also have `lifecycle`. This should not be an issue.

Reply:
> We'd have to remove every version of each file, so I hope to minimize the scope of the impact as much as possible.

Please don't overestimate the *impact* of test artifacts. In the worst cases, we have some files left in a bucket which is fine.

> Since version control should not affect normal user usage

`version` is part of our public API. Please don't think of it's *not affect normal user usage*. It should be treated the same as `range`, `buffer`, `concurrent`. I see no reason to test `version` separately.

> there should be no need to test the functionality of version control unless it is enabled.

This is why we have `capability`.  

---


> 1. Add version-related test cases to the existing behavior test. In this way, we can use environment variables to skip cases that do not involve version control, based on whether version control is enabled.

We don't require additional environment variables. Instead, we need a fresh setup for services. Take S3, for instance; we require a newly configured bucket with the feature enabled. 

Additionally, we must introduce a new flag for the S3 service, which should be integrated into our API rather than configured in CI.

Reply:
Wow, that's a great news.
We could keep them for just 1 day.
Could the test bucket we're currently using be configured the same way?

Reply:
> We could keep them for just 1 day.
> Could the test bucket we're currently using be configured the same way?

Yes, already have.

Reply:
> I'm mainly considering the cleanup issue. With storage services that have version control, cleanup can be more complicated. We'd have to remove every version of each file, so I hope to minimize the scope of the impact as much as possible. Since version control should not affect normal user usage, there should be no need to test the functionality of version control unless it is enabled.
> 
> So, can we:
> 
> 1. Add version-related test cases to the existing behavior test. In this way, we can use environment variables to skip cases that do not involve version control, based on whether version control is enabled. But it also needs to add some extra cleanup steps.
> 2. Create a separate behavior test suite(maybe like this PR, but not in edge tests) that does not affect existing tests.

The test infra would not be a problem we just need to configure the lifecycle like the following below

![image](https://github.com/apache/incubator-opendal/assets/7054676/f4c35d12-3b6c-4ac6-8b38-698efe445ec1)